### ISSUE TITLE
`UpdateCacheOp` requires the cache operand to be exactly rank-4

### ISSUE THREAD

**Issue by elogir** [tenstorrent/tt-mlir#8619]
## Summary

`UpdateCacheOp`'s verifier rejects any cache shape that isn't exactly rank-4. The shape is hardcoded in the verifier rather than inferred from the operands.

## Source

https://github.com/tenstorrent/tt-mlir/blob/f400e65a698eb4a5385be94397cfa4ff58f7443d/lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp#L6263-L6265
https://github.com/tenstorrent/tt-mlir/blob/f400e65a698eb4a5385be94397cfa4ff58f7443d/lib/Dialect/TTIR/IR/TTIROps.cpp#L5721-L5727

Same rank-4 check for `FillCacheOp` / paged variants.

## Example

A shared KV cache representation like a single rank-5 tensor `[layer, batch, head, position, head_dim]` with the layer dim flowing through to the cache update never lowers.

## Possible resolution

The StableHLO matcher should not reject rank-5 scatter inputs that are otherwise compatible with cache semantics. Letting frontends ship a single shared cache avoids inflating the entry-point signature with N independent per-layer tensors and unlocks more natural sharding layouts.

Tested with `pjrt-plugin-tt==1.2.0.dev20260526002843`

**Comment by sdjukicTT** [comment]
Currently there are no kernels that accept rank 5 that compiler could use directly, we are discussing this with our kernel team on [issue](https://github.com/tenstorrent/tt-metal/issues/46249).

Potentionally we could make a workaround in compiler for this (which would be faster), but we need more info:
- where do you guys use this?
- how does input to our compiler look like (it would be awesome if you could attach a repro or StableHLO IR)?
- do you need paged variant of `fill`/`update_cache` or would regular variant be enough?

**Comment by elogir** [comment]
We usually use it like that:

https://github.com/zml/zml/blob/c5bdcac63205c9ce27c3703edcb26508f1d81dca/examples/llm/models/qwen3_5/model.zig#L973-L979
https://github.com/zml/zml/blob/c5bdcac63205c9ce27c3703edcb26508f1d81dca/examples/llm/models/qwen3_5/model.zig#L1016-L1020

i.e. the whole KV cache is a single shared rank-5 tensor `[.b, .layer, .s, .h, .hd]`.

We usually emit a `dynamic_update_slice`; the `scatter` was a workaround for the now-fixed https://github.com/tenstorrent/tt-mlir/issues/8618. Here are both versions, with a jax reproducer:

```
// ===== dus : StableHLO in =====
module @jit_dus attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<1x8x64x4x256xbf16>, %arg1: tensor<1x1x64x4x256xbf16>, %arg2: tensor<ui32>) -> (tensor<1x8x64x4x256xbf16> {jax.result_info = "result"}) {
    %c = stablehlo.constant dense<0> : tensor<ui32>
    %0 = stablehlo.dynamic_update_slice %arg0, %arg1, %c, %c, %arg2, %c, %c : (tensor<1x8x64x4x256xbf16>, tensor<1x1x64x4x256xbf16>, tensor<ui32>, tensor<ui32>, tensor<ui32>, tensor<ui32>, tensor<ui32>) -> tensor<1x8x64x4x256xbf16>
    return %0 : tensor<1x8x64x4x256xbf16>
  }
}

// ===== scatter : StableHLO in =====
module @jit_scatter attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<1x8x64x4x256xbf16>, %arg1: tensor<1x2xui32>, %arg2: tensor<1x1x4x256xbf16>) -> (tensor<1x8x64x4x256xbf16> {jax.result_info = "result"}) {
    %0 = "stablehlo.scatter"(%arg0, %arg1, %arg2) <{indices_are_sorted = true, scatter_dimension_numbers = #stablehlo.scatter<update_window_dims = [0, 2, 3], inserted_window_dims = [1, 2], scatter_dims_to_operand_dims = [1, 2], index_vector_dim = 1>, unique_indices = false}> ({
    ^bb0(%arg3: tensor<bf16>, %arg4: tensor<bf16>):
      stablehlo.return %arg4 : tensor<bf16>
    }) : (tensor<1x8x64x4x256xbf16>, tensor<1x2xui32>, tensor<1x1x4x256xbf16>) -> tensor<1x8x64x4x256xbf16>
    return %0 : tensor<1x8x64x4x256xbf16>
  }
}
```

```python
import os, glob, shutil, tempfile
os.environ["TT_METAL_LOGGER_LEVEL"] = "FATAL"
import jax, jax.numpy as jnp
from jax import lax
jax.config.update("jax_platforms", "tt")

B, L, S, H, D = 1, 8, 64, 4, 256
U = jnp.uint32

def dus(cache, new, pos):
    z = U(0)
    return lax.dynamic_update_slice(cache, new, (z, z, pos, z, z))

dnums = lax.ScatterDimensionNumbers(
    update_window_dims=(0, 2, 3), inserted_window_dims=(1, 2),
    scatter_dims_to_operand_dims=(1, 2))

def scatter(cache, idx, new):
    return lax.scatter(cache, idx, new, dnums, indices_are_sorted=True, unique_indices=False)

def show(name, fn, *args):
    print(f"\n// ===== {name} : StableHLO in =====")
    print(jax.jit(fn).lower(*args).as_text())
    d = tempfile.mkdtemp()
    try:
        jax.jit(fn, compiler_options={"export_path": d, "export_model_name": name}).lower(*args).compile()
        print(f"// ===== {name} : TTIR out =====")
        for f in sorted(glob.glob(f"{d}/**/ttir_*{name}*.mlir", recursive=True)):
            for line in open(f):
                if any(o in line for o in ("update_cache", "fill_cache", "ttir.scatter",
                                           "dynamic_update_slice", "cpu_hoisted")):
                    print("  " + line.strip())
    finally:
        shutil.rmtree(d, ignore_errors=True)

show("dus", dus, jnp.zeros((B, L, S, H, D), jnp.bfloat16),
     jnp.ones((B, 1, S, H, D), jnp.bfloat16), U(0))
show("scatter", scatter, jnp.zeros((B, L, S, H, D), jnp.bfloat16),
     jnp.array([[0, 5]], jnp.uint32), jnp.ones((B, 1, H, D), jnp.bfloat16))
```

Neither reaches `update_cache` / `fill_cache` over the shared rank-5 cache (the reproducer prints the TTIR for both): the `dynamic_update_slice` is hoisted to the host as a `ttir.cpu_hoisted_call`, and the `scatter` falls to a generic `ttir.scatter` over the cache flattened to 1-D (`tensor<524288xbf16>`).

Regular `update_cache` / `fill_cache` is what unblocks the models we run on TT today as an example but a paged variant would be great, as our production inference server is paged-native, so we'll need it there too.

**Comment by sdjukicTT** [comment]
Great, thank you for the detailed response and example!
We will look into how this can be supported.
